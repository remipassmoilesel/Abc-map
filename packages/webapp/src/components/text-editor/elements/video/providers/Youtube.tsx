/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { Provider } from '../Provider';
import { DefaultIntegration } from '../DefaultIntegration';

const shortUrlRegexp = /youtu\.be\/([0-9A-Za-z_-]+)/i;
const longUrlRegexp = /youtube\.com\/watch\?v=([0-9A-Za-z_-]+)/i;

export class Youtube implements Provider {
  public canHandle(url: string): boolean {
    return !!this.extractVideoId(url);
  }

  public getIntegration(url: string): React.ReactNode {
    const videoId = this.extractVideoId(url);
    if (!videoId) {
      return <DefaultIntegration url={url} />;
    }

    return (
      <iframe
        title="YouTube video player"
        src={`https://www.youtube.com/embed/${videoId}`}
        frameBorder="0"
        allow="accelerometer; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
        allowFullScreen
      />
    );
  }

  private extractVideoId(url: string): string | undefined {
    const match = url.match(shortUrlRegexp) || url.match(longUrlRegexp);
    if (!match || !match.length) {
      return;
    }

    return match[1];
  }
}
