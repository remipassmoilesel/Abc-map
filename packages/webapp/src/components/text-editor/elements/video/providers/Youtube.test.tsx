/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Youtube } from './Youtube';
import { render } from '@testing-library/react';

describe('Youtube', () => {
  let provider: Youtube;
  beforeEach(() => {
    provider = new Youtube();
  });

  it('should handle', () => {
    expect(provider.canHandle('https://youtu.be/uTC1mW4ua74')).toBe(true);
    expect(provider.canHandle('https://www.youtube.com/watch?v=uTC1mW4ua74')).toBe(true);
    expect(provider.canHandle('https://docs.slatejs.org/api/nodes')).toBe(false);
  });

  it('should return integration for short url', () => {
    const { container } = render(<div>{provider.getIntegration('https://youtu.be/uTC1mW4ua74')}</div>);
    expect(container).toMatchSnapshot();
  });

  it('should return integration for long url', () => {
    const { container } = render(<div>{provider.getIntegration('https://www.youtube.com/watch?v=uTC1mW4ua74')}</div>);
    expect(container).toMatchSnapshot();
  });

  it('should work with https://youtu.be/tEpMEl-_syY', () => {
    const { container } = render(<div>{provider.getIntegration('https://youtu.be/tEpMEl-_syY')}</div>);
    expect(container).toMatchSnapshot();
  });
});
