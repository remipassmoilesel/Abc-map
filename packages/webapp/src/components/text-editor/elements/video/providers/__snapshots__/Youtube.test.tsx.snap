// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Youtube should return integration for long url 1`] = `
<div>
  <div>
    <iframe
      allow="accelerometer; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
      allowfullscreen=""
      frameborder="0"
      src="https://www.youtube.com/embed/uTC1mW4ua74"
      title="YouTube video player"
    />
  </div>
</div>
`;

exports[`Youtube should return integration for short url 1`] = `
<div>
  <div>
    <iframe
      allow="accelerometer; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
      allowfullscreen=""
      frameborder="0"
      src="https://www.youtube.com/embed/uTC1mW4ua74"
      title="YouTube video player"
    />
  </div>
</div>
`;

exports[`Youtube should work with https://youtu.be/tEpMEl-_syY 1`] = `
<div>
  <div>
    <iframe
      allow="accelerometer; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
      allowfullscreen=""
      frameborder="0"
      src="https://www.youtube.com/embed/tEpMEl-_syY"
      title="YouTube video player"
    />
  </div>
</div>
`;
